### ISSUE TITLE
Audio playback interruption on Android devices

### ISSUE THREAD

**Issue by yelkamel** [ryanheise/just_audio#1363]
**Which API doesn't behave as documented, and how does it misbehave?**
The just_audio package for Flutter is not behaving as expected on Android devices. Specifically, the audio playback is unexpectedly stopping for some users, even though the progress bar continues to move. This behavior is not documented and appears to be a bug.

**Minimal reproduction project**
Unfortunately, we cannot provide a minimal reproduction project at this time as the issue is not consistently reproducible in development or release environments. It only occurs for some users in production.

**To Reproduce (i.e. user steps, not code)**
1. Start playing an audio file (meditation or sleep music) using the just_audio package
2. Let the audio play for an indeterminate amount of time
3. Audio playback stops unexpectedly, but the progress bar continues to move
4. Attempting to restart playback results in the progress bar moving without audio

**Error messages**
We don't have any error in crashanalytics.

**Expected behavior**
The audio should play continuously without interruption for the entire duration of the audio file.

**Screenshots**
No screenshots are available at this time as the issue occurs randomly for users.

**Desktop (please complete the following information):**
N/A - This issue is specific to Android devices.

**Smartphone (please complete the following information):**
 - Device: Various Android devices
 - OS: Various Android versions

**Flutter SDK version**
- Flutter version: (sdk: '>=3.0.0 <4.0.0')
-  just_audio: ^0.9.35
- audio_service: ^0.18.15

**Additional context**
- This issue only affects some Android users and is not reproducible in our development or release environments.
- The problem started occurring recently, but the app was working fine before.
- We've verified it's not a memory leak issue.
- Affected users seem to have sufficient RAM available (based on crash analytics data).
- No crashes are reported in our crash analytics related to this issue.
- The problem occurs for both short and long audio files.
- The issue persists whether the screen is locked or unlocked.
- It happens both when the app is in the foreground and background.
- Some users report that restarting the audio from the beginning after the interruption allows it to play normally.

We're seeking assistance in identifying potential causes or additional debugging steps we can take to resolve this issue.

**Comment by Chaphasilor** [comment]
We are seeing similar reports for Finamp. Related issue: https://github.com/jmshrv/finamp/issues/956  
Users report playback stopping a few seconds or minutes after leaving the app or locking the screen (which could be exactly the buffer duration).

**Comment by srihariash999** [comment]
This happened to me a few times recently and two times out of those the audio cut off suspiciously soon after I recieved a notification. Not sure if that really is something or just a coincidence but leaving this comment here anyway in case somebody else also observes it.

**Comment by mklepaczko** [comment]
Same here, player is not stopped, progress continues but there is no sound. Some users reported it happens after 5 - 6 mins. Perhaps a new energy saving feature on Android.

**Comment by srihariash999** [comment]
@mklepaczko Now that you mentioned it, I am only seeing this on my pixel 8 after android 15 update and latest patch and the timing almost seems correct too. 

Can you mention which android version you are seeing this issue in?  Also it'll be great if someone can test and verify if this issue is present for older package version or android versions or combo of both.

**Comment by ryanheise** [comment]
Is anyone able to submit a minimal reproduction project to allow me to investigate?

I have a Pixel 8, so I ought to be able to reproduce it if someone can provide a reproduction project. Before you do so, of course please make sure your issue either happens in the foreground, or alternatively if this is a background thing, ensure that background audio is properly configured to keep it alive.

**Comment by marianperca** [comment]
Hi, I had the ~ same issue, streaming stopped when the device display was closed and nothing was shown in the notification tray. The fix for me was adding the new foreground service type declaration required starting with android 14: https://developer.android.com/about/versions/14/changes/fgs-types-required
```
<uses-permission android:name="android.permission.FOREGROUND_SERVICE" />
<uses-permission android:name="android.permission.FOREGROUND_SERVICE_MEDIA_PLAYBACK" />
```

**Comment by Chaphasilor** [comment]
@marianperca good catch! We only have the issue on the stable version, not the beta version, and there really is a difference in permissions. Didn't see that in previous diffs since the branches have diverged a lot in our case.  
We'll push a fix with the new permissions soon and report back!

**Comment by umbertogrimaldi** [comment]
@marianperca Unfortunately for me it's not the solution. I have these 2 permissions in the manifest implemented since 10 months. My attempt to fix this was to set `androidStopForegroundOnPause` to false in the `AudioService` configuration, but I'm not sure if that's the solution or not.

<img width="828" alt="Screenshot 2024-12-27 at 09 34 42" src="https://github.com/user-attachments/assets/1b15fe94-a778-4042-bfaa-30be13e92459" />

**Comment by MatteoBax** [comment]
Same issue with background playback. 
This issue also occurs from browsers.

**Comment by johnmollaghan** [comment]
I am seeing this issue also on a Samsung A52 (Android 14). I am in the process of writing a Podcast app and this is a major blocker for me I'm afraid.

I am testing using a long audio stream (meditation podcast). There seem to be TWO problems that can happen that I have seen.

- Even if I don't touch the phone and just let the audio play (app background, phone locked), it cannot get passed around the 40 minute mark.
- It also seems to have issued  6 or 7 minutes AFTER getting a phone call.

The app doesn't crash, the sound just stops BUT the audio player looks like it it still playing.

Could moving to just_audio_background resolve this? I really don't want to as I want to have custom notification player buttons.

All help welcome, but I think I am following the recommended implementation.

I also tried this on a Samsung A9+ Tablet running Android 14 and the Tablet seemed to work OK, the audio played beyond the hour mark etc.

**Comment by Chaphasilor** [comment]
> We'll push a fix with the new permissions soon and report back!

The issue appears to have been fixes by properly requesting all needed permissions in our case. I'd encourage everyone to take a close look at the just_audio readme and make sure everything is configured as laid out there :)
